woocommerce/woocommerce-ios
# Login: Remove XMLRPC dependency for the native Jetpack Installation flow

Closes: #8258 
Please also review https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/711.



### Description

Currently, we are using XMLRPC API to check the input username & password and get the URLs for login and admin pages, which is then used to construct `CookieNonceAuthenticator` for handling the authentication for REST APIs. If a site blocks access to XMLRPC, requests to its API will fail, and the authenticator cannot be created.

This PR fixes the above issue by removing the call to XMLRPC API. Instead, we are handling authentication with the following changes:
- Remove the XMLRPC check for site discovery if native Jetpack is enabled.
- Guess the login and admin URLs based on the site URL.
- Construct `CookieNonceAuthenticator` based on the guessed URLs and hardcoded WP version for Ajax nonce retrieval.
- Attempt fetching Jetpack plugin details:
  - If the request succeeds or gets a 404 or 403 error, the API can be reached. Proceed to trigger to completion closure.
  - If the request gets a 401 error, the authentication fails. An alert will be displayed with a message about incorrect credentials.
  - A generic error is displayed if the request fails with any other error.
- Remove XMLRPC from the completion callback for the site credential login screen. This should not be required for verifying a WP.com account. 

### Testing instructions

Make sure you have access to a test site that doesn't have Jetpack.
- Since the native Jetpack installation flow is A/B tested, disable [the check](https://github.com/woocommerce/woocommerce-ios/blob/a1635986f6d3b6399d7bb879c1cb1fa77a2b4bc2/WooCommerce/Classes/Authentication/AuthenticationManager.swift#L764) to return Jetpack setup UI by default. Make sure to set the config `skipXMLRPCCheckForSiteDiscovery` to true in `AuthenticationManager` too. 
- Build the app and proceed to log in with a WP.com account.
- On the store picker, select Add a site > Connect to an existing site.
- Enter the address of your test site. Notice in the Xcode console that there is no message regarding any XMLRPC check.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- On the site credential screen, enter incorrect credentials. Notice that an alert is displayed for incorrect credentials.
- Try again with the credentials of an account that doesn't have permission to manage plugins. Notice that the setup screen is displayed, and the permission error appears.
- Try again with the correct credentials, notice that the Jetpack Setup screen is displayed.
- On the connection web view, select Sign in with a different account.
- Log in with a WP.com account that is different from your logged-in once and authorize the connection.
- When the setup is complete, select Go to Store.
- Notice that a new screen is presented to verify the new WP.com email. Enter the password for that account, notice that the login succeeds, and you are navigated to the home screen.

### Screenshots


https://user-images.githubusercontent.com/5533851/204742152-0861fb42-c773-495b-8a17-9ab3aec85fb2.mp4



---
